jtoh.go

```diff
@@ -1,7 +1,157 @@
-package main
+package jtoh
 
-import "fmt"
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"strings"
+)
 
-func main() {
-	fmt.Println("TODO")
+// J is a jtoh transformer, it transforms JSON into something more human
+type J struct {
+	separator      string
+	fieldSelectors []string
+}
+
+// Err is a jtoh error
+type Err string
+
+const InvalidSelectorErr Err = "invalid selector"
+
+// New creates a new jtoh transformer using the given selector.
+// The selector is on the form <separator><field selector 1><separator><field selector 2>
+// For example, given ":" as a separator you can define:
+//
+// :fieldA:fieldB:fieldC
+//
+// Accessing a nested field is done with dot notation, like this:
+//
+// :field.nested
+//
+// Making "." the only character that will not be allowed to be used
+// as a separator since it is already a selector for nested fields.
+//
+// If the selector is invalid it returns an error.
+func New(s string) (J, error) {
+	selector := []rune(s)
+	if len(selector) <= 1 {
+		return J{}, fmt.Errorf("%w:%s", InvalidSelectorErr, s)
+	}
+	separator := string(selector[0])
+	return J{
+		separator:      separator,
+		fieldSelectors: trimSpaces(strings.Split(string(selector[1:]), separator)),
+	}, nil
+}
+
+// Do receives a json stream as input and transforms it
+// in simple lines of text (newline-delimited) which is
+// then written in the provided writer.
+//
+// This function will block until all data is read from the input
+// and written on the output.
+func (j J) Do(jsonInput io.Reader, textOutput io.Writer) {
+	jsonInput, ok := isList(jsonInput)
+	dec := json.NewDecoder(jsonInput)
+
+	if ok {
+		// WHY: To handle properly gigantic lists of JSON objs
+		// Really don't need the return value, but linters can be annoying =P
+		_, _ = dec.Token()
+	}
+
+	for dec.More() {
+		m := map[string]interface{}{}
+		err := dec.Decode(&m)
+		if err != nil {
+			// TODO: handle non disruptive parse errors
+			// Ideally we want the original non-JSON data
+			// Will need some form of extended reader that remembers
+			// part of the read data (not all, don't want O(N) spatial
+			// complexity).
+			fmt.Fprintf(textOutput, "TODO:HANDLERR:%v\n", err)
+			return
```

Comment:
Descriptions says: "This function will block until all data is read from the input and written on the output"
But since we have this `return` here, it can also be "until there is an error when parsing"

Reply:
I haven't fixed this on the docs because the idea is to not stop when there is an error when parsing, I just haven't implemented that yet =P (like I started with the interface/behavior I want to have, but I'm not there yet)